### PR TITLE
Fix/challenge for bbr

### LINKED ISSUE #673: Challenges are not verified
The challenges are not being verified and validated by Blobbers and validators due to the API changes of `/openchallenges` and `/getchallenge` in 0chain repo. We need to fix this ASAP.

### PR DESCRIPTION
**Pull request by cnlangzi**
### Changes
-

### Fixes
- fixed UpdateBlobberOnChain issue 
- reverted  GetBlocksMerkleTreeForChallenge to fix challenge building. it is broken by https://github.com/0chain/blobber/pull/642. 
- fixed cli.sh that is broken by https://github.com/0chain/blobber/pull/642

### Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/blobber/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

### Associated PRs (Link as appropriate):
-

### CODE REVIEW

File: code/go/0chain.net/blobbercore/filestore/storage.go [hunk 1]
```diff
@@ -385,32 +387,51 @@ func (fs *FileStore) GetBlocksMerkleTreeForChallenge(allocID string,
...
-		if err != nil {
-			return nil, nil, common.NewError("buffer_write_error", err.Error())
-		}
+			merkleChunkSize := int(fileData.ChunkSize) / 1024
 
-		startInd := blockoffset * MerkleChunkSize
-		endInd := startInd + MerkleChunkSize
-		returnBytes = append(returnBytes, dataBytes[startInd:endInd]...)
+			if merkleChunkSize == 0 {
+				merkleChunkSize = 1
```
Comment (peterlimg) [inline code/go/0chain.net/blobbercore/filestore/storage.go]: The merkleChunSize can be calculated outside of the for loop, if the fileData.ChunSize is always 64K, then we can use the const variable MerkleChunkSize directly.

File: code/go/0chain.net/blobbercore/filestore/storage.go [hunk 2]
```diff
@@ -385,32 +387,51 @@ func (fs *FileStore) GetBlocksMerkleTreeForChallenge(allocID string,
...
 	fixedMT := util.NewFixedMerkleTree(int(fileData.ChunkSize))
 
 	bytesBuf := bytes.NewBuffer(make([]byte, 0))
-	for i := 0; i < totalChunks; i++ {
-		n, err := io.CopyN(bytesBuf, file, fileData.ChunkSize)
-		if err != nil {
-			if n == 0 && errors.Is(err, io.EOF) {
-				break
+	for chunkIndex := 0; chunkIndex < numChunks; chunkIndex++ {
+		written, err := io.CopyN(bytesBuf, file, fileData.ChunkSize)
```
Comment (peterlimg) [inline code/go/0chain.net/blobbercore/filestore/storage.go:398]: Check error first like before to make the code clean.

File: code/go/0chain.net/blobbercore/filestore/storage.go [hunk 3]
```diff
@@ -385,32 +387,51 @@ func (fs *FileStore) GetBlocksMerkleTreeForChallenge(allocID string,
...
+
+		if written > 0 {
+			dataBytes := bytesBuf.Bytes()
+
+			errWrite := fixedMT.Write(dataBytes, chunkIndex)
+			if errWrite != nil {
+				return nil, nil, thrown.ThrowLog(errWrite.Error(), constants.ErrUnableHash)
```
Comment (peterlimg) [inline code/go/0chain.net/blobbercore/filestore/storage.go]: Return error like before to make the code clean unless the ErrUnableHash is a must

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/blobbercore/filestore/storage.go [hunk 4]
```diff
@@ -385,32 +385,50 @@ func (fs *FileStore) GetBlocksMerkleTreeForChallenge(allocID string,
...
 			}
-			return nil, nil, err
-		}
 
-		dataBytes := bytesBuf.Bytes()
-		err = fixedMT.Write(dataBytes, i)
-		if err != nil {
-			return nil, nil, common.NewError("buffer_write_error", err.Error())
-		}
+			if merkleChunkSize == 0 {
```
Comment (peterlimg) [inline code/go/0chain.net/blobbercore/filestore/storage.go]: This merkleChunkSize checking can also be moved out of the loop

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/blobbercore/filestore/storage.go (hunk 4 above)
Comment (cnlangzi) [reply code/go/0chain.net/blobbercore/filestore/storage.go]: fixed

**Comment by lpoli** [comment]
@cnlangzi I don't understand what was the fix by this PR. 
I also had provided single block and not all data. I calculated pre-hand for which you checked with for loop and incrementing offset.

I've considered that chunk size will always be 64KB. If you are deviating from 64KB and fixing for other chunk sizes then we have created a complete mess.

Also there existed no `mount_point` issue.

**Comment by peterlimg** [comment]
@lpoli the previous logic would get block every 1024bytes

**Review by peterlimg — APPROVED** [review]
lgtm

**Comment by lpoli** [comment]
> @lpoli the previous logic would get block every 1024bytes
```
startInd := blockoffset * MerkleChunkSize
endInd := startInd + MerkleChunkSize
returnBytes = append(returnBytes, dataBytes[startInd:endInd]...)
```

Above was the logic, I don't see how it would get every 1024 blocks.

**Pull request merged**